core/services/workflows/engine.go

```diff
@@ -148,73 +154,96 @@ func (e *Engine) triggerHandlerLoop(ctx context.Context) {
-				"observations": &values.List{
-					Underlying: []values.Value{event.Value},
+	ec := &executionState{
+		steps: map[string]*stepState{
+			trigger.Ref: {
+				outputs: &stepOutput{
+					value: event.Value,
 				},
 			},
 		},
-		Config: e.consensusConfig,
+		workflowID:  mockedWorkflowID,
+		executionID: mockedExecutionID,
 	}
-	chReports := make(chan capabilities.CapabilityResponse, 10)
-	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	err := e.consensus.Execute(newCtx, chReports, cr)
+
+	consensus := e.workflow.Consensus[0]
+	err := e.handleStep(ctx, ec, consensus)
 	if err != nil {
-		return nil, err
+		e.logger.Errorf("error in handleConsensus %v", err)
+		return
 	}
-	select {
-	case resp := <-chReports:
-		if resp.Err != nil {
-			return nil, resp.Err
+
+	for _, trg := range e.workflow.Targets {
+		err := e.handleStep(ctx, ec, trg)
+		if err != nil {
+			e.logger.Errorf("error in handleTargets %v", err)
+			return
 		}
-		return resp.Value, nil
-	case <-ctx.Done():
-		return nil, ctx.Err()
 	}
 }
 
-func (e *Engine) handleTargets(ctx context.Context, resp values.Value) error {
-	e.logger.Debugw("handle targets")
-	inputs := map[string]values.Value{
-		"report": resp,
+func (e *Engine) handleStep(ctx context.Context, es *executionState, node Capability) error {
+	stepState := &stepState{
+		outputs: &stepOutput{},
 	}
+	es.steps[node.Ref] = stepState
 
-	var combinedErr error
-	for _, t := range e.targets {
-		e.logger.Debugw("sending to target", "target", t.typeStr, "inputs", inputs)
-		tr := capabilities.CapabilityRequest{
-			Inputs: &values.Map{Underlying: inputs},
-			Config: t.config,
-			Metadata: capabilities.RequestMetadata{
-				WorkflowID:          mockedWorkflowID,
-				WorkflowExecutionID: mockedExecutionID,
-			},
-		}
-		_, err := capabilities.ExecuteSync(ctx, t.capability, tr)
-		combinedErr = errors.Join(combinedErr, err)
+	// Let's get the capability. If we fail here, we'll bail out
+	// and try to handle it at the next execution.
+	cp, err := e.registry.Get(ctx, node.Type)
+	if err != nil {
+		return err
+	}
+
+	api, ok := cp.(capabilities.CallbackExecutable)
+	if !ok {
+		return fmt.Errorf("capability %s must be an action, consensus or target", node.Type)
 	}
-	return combinedErr
+
+	i, err := findAndInterpolateAllKeys(node.Inputs, es)
+	if err != nil {
+		return err
+	}
+
+	inputs, err := values.NewMap(i.(map[string]any))
+	if err != nil {
+		return err
+	}
+
+	stepState.inputs = inputs
+
+	config, err := values.NewMap(node.Config)
+	if err != nil {
+		return err
+	}
+
+	tr := capabilities.CapabilityRequest{
+		Inputs: inputs,
+		Config: config,
+		Metadata: capabilities.RequestMetadata{
+			WorkflowID:          es.workflowID,
+			WorkflowExecutionID: es.executionID,
+		},
+	}
+
+	resp, err := capabilities.ExecuteSync(ctx, api, tr)
+	if err != nil {
+		stepState.outputs.err = err
+		return err
+	}
+
+	if len(resp.Underlying) > 1 {
```

Comment:
the meaning of this is not obvious to untrained eyes. consider a comment or well named helper

Reply:
Good point! I'll add a comment